### DISCUSSION TITLE
Create a model to automatically determine the message type

### DISCUSSION THREAD

**Discussion by HomerusJa** [pydantic/pydantic#10672] [Question]
I've got the following usecase of pydantic:

I get data from an API which I want to parse. Following, a stripped down version of the pydantic models I defined describing the messages:

```python
from pydantic import BaseModel, Field, RootModel
from typing import TypeVar, Generic, Literal, Union

T = TypeVar('T')

class S3IMessageBodyBase(BaseModel):
    sender: str
    identifier: str
    receivers: list[str]
    messageType: str

class GetValueRequestBody(S3IMessageBodyBase):
    messageType: Literal["getValueRequest"]
    replyToEndpoint: str
    attributePath: str

class GetValueReplyBody(S3IMessageBodyBase, Generic[T]):
    messageType: Literal["getValueReply"]
    replyingToMessage: str

    # The library should not know about the different values being passed in, but those should also be able to be validated
    value: T

# Many (about 20) more request and reply types
```

Now I want to have a pydantic model which I can just initialize with the content of the response and which then determines the messageType and so on for me. I was thinking about a usage like the following:

```python
message_content: dict = broker.receive_message()
message = s3i.models.S3IMessage(message_content)
```

I am really new to pydantic (in fact, this is my second use of it), and I've searched the docs for solutions, but could not find one. I found the [RootModel](https://docs.pydantic.dev/2.3/usage/models/#rootmodel-and-custom-root-types) which might be helpful, but I dont really understand how to use it correctly.

**Comment by HomerusJa** [comment]
For reference, a typical GetValueReplyBody looks like the following:

```json
{
  "sender": "s3i:ef39a0ae-1f4a-4393-9508-ad70a4d38a63",
  "identifier": "s3i:e89a4538-19fd-4675-8e55-9a7e59885c85",
  "receivers": [
    "s3i:2073c475-fee5-463d-bce1-f702bb06f899"
  ],
  "messageType": "getValueReply",
  "replyingToMessage": "s3i:147ecc53-4f5e-4270-9672-0174a017be7b",
  "value": {"key": "value"}
}
```

**Comment by ashb** [comment]
```python
class GetValueRequestBody(S3IMessageBodyBase):
    messageType: Literal["getValueRequest"] = "getValueRequest"
    replyToEndpoint: str
    attributePath: str

class GetValueReplyBody(S3IMessageBodyBase, Generic[T]):
    messageType: Literal["getValueReply"] = "getValueReply"
    replyingToMessage: str
```

I don't know/can't find any automatic way of doing it, but this means you don't have to pass the type when creating your body objects.

**Reply by HomerusJa** [reply]
Thanks for your answer :pray: 

Still, I don’t really understand your solution. What did you change? Isn’t the code above just a slice from my example?
